File: Zamba/ZambaWeb2013/zamba2010/Zamba.Web2014/Views/UC/WF/Arbol.ascx.vb
Imports Zamba.Core

Partial Public Class Arbol
    Inherits System.Web.UI.UserControl
    Implements IArbol

    ''' <summary>
    ''' Una vez finalizado el refresco del arbol de wf se eleva este evento para actualizar la grilla
    ''' </summary>
    ''' <param name="StepId"></param>
    ''' <remarks></remarks>
    Public Event WFTreeRefreshed(ByVal StepId As Int32)
    Public Event WFTreeIsEmpty()
    Public Event SelectedNodeChanged(ByVal WFId As Int32, ByVal StepId As Int32, ByVal DocTypeId As Int32)
    Private Presenter As Presenters.Arbol

    Public Sub Arbol()
    End Sub

    Protected Sub Page_Init(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Init
        If Not IsNothing(Session("User")) Then
            Presenter = New Presenters.Arbol(Session("UserId"), Me, Session("User"))
        End If
    End Sub

    Protected Sub Page_Load(sender As Object, e As EventArgs) Handles Me.Load
        If Not IsNothing(Session("User")) Then
            btnInsertar.Visible = UserBusiness.Rights.GetUserRights(Zamba.Core.ObjectTypes.InsertWeb, Zamba.Core.RightsType.View)
            If btnInsertar.Visible Then
                Presenter.FillCmbFormTypes()
            End If
        End If

        Dim script As String = "$(document).ready(function() { FixWfTree(); AddStepCountHandler(); });"
        ScriptManager.RegisterClientScriptBlock(Me, Me.[GetType](), "WfTreeScripts", script, True)
    End Sub

    Public Sub Refresh()
        Try
            Dim LastWFStepUsed As String = UserPreferences.getValue("UltimoWFStepUtilizado", Sections.WorkFlow, "")
            Dim WFId As Int32
            Dim stepID As Int32

            If String.IsNullOrEmpty(LastWFStepUsed) = False Then
                Presenter.getWorkflowAndStepSelected(WFId, LastWFStepUsed)
                SelectStepNode(LastWFStepUsed)

                If ArbolProcesos.Nodes.Count > 0 Then
                    Dim nodeselected As Boolean = False
                    For Each nodeMain As TreeNode In ArbolProcesos.Nodes(0).ChildNodes

                        If nodeMain.Value = LastWFStepUsed Then
                            nodeMain.Select()
                            RaiseEvent SelectedNodeChanged(WFId, LastWFStepUsed, 0)
                            Exit For
                        End If

                        For Each NodeChild As TreeNode In nodeMain.ChildNodes
                            If NodeChild.Value = LastWFStepUsed Then
                                nodeMain.Expand()
                                NodeChild.Select()
                                RaiseEvent SelectedNodeChanged(WFId, LastWFStepUsed, 0)
                                nodeselected = True
                                Exit For
                            End If
                        Next
                        If nodeselected Then Exit For
                    Next



                End If
            Else
                Presenter.getWorkflowAndStepSelected(WFId, stepID)
                RaiseEvent SelectedNodeChanged(WFId, stepID, 0)
            End If
        Catch ex As Exception
            ZCore.raiseerror(ex)
        End Try
    End Sub

    Private Sub RefreshTree()
        Try
            Presenter.FillWF()

            Dim LastWFStepUsed As String = UserPreferences.getValue("UltimoWFStepUtilizado", Sections.WorkFlow, "")
            Dim WFId As Int32

            Presenter.getWorkflowAndStepSelected(WFId, LastWFStepUsed)

            If ArbolProcesos.Nodes.Count > 0 Then
                SelectStepNode(LastWFStepUsed)
                RaiseEvent WFTreeRefreshed(LastWFStepUsed)
            End If
        Catch ex As Exception
            ZCore.raiseerror(ex)
        End Try
    End Sub

    Private Sub SelectStepNode(ByVal stepId As Int64)
        ArbolProcesos.Nodes(0).Expand()

        For Each nodeMain As TreeNode In ArbolProcesos.Nodes(0).ChildNodes

            If nodeMain.Value = stepId Then
                nodeMain.Select()
            End If

            For Each NodeChild As TreeNode In nodeMain.ChildNodes
                If NodeChild.Value = stepId Then
                    nodeMain.Expand()
                    NodeChild.Select()
                End If
            Next
        Next
    End Sub

    Public Sub FillWF()
        Try
            Presenter.FillWF()

            'Si el arbol no posee nodos, se dispara el evento para que no muestre el control
            If ArbolProcesos.Nodes.Count = 0 Then
                RaiseEvent WFTreeIsEmpty()
                Return
            End If

            If IsNothing(ArbolProcesos.SelectedNode) Then
                Refresh()
            End If
        Catch ex As Exception
            ZCore.raiseerror(ex)
        End Try
    End Sub

    Public Property WFTreeView() As TreeView Implements IArbol.WFTreeView
        Get
            Return Me.ArbolProcesos
        End Get
        Set(ByVal value As TreeView)
            Me.ArbolProcesos = value
        End Set
    End Property

    'Protected Sub btncontraer_Click(ByVal sender As Object, ByVal e As System.Web.UI.ImageClickEventArgs) Handles btnContraer.Click
    '    Presenter.CollapseTreeView()
    'End Sub

    Public Property CmbFormType() As System.Web.UI.WebControls.DropDownList Implements IArbol.CmbFormType
        Get
            Return Me.ddltipodeform
        End Get
        Set(ByVal value As System.Web.UI.WebControls.DropDownList)
            Me.ddltipodeform = value
        End Set
    End Property

    Protected Sub BtnRefresh_Click(ByVal sender As Object, ByVal e As EventArgs)
        RefreshTree()
    End Sub

    Protected Sub ArbolProcesos_SelectedNodeChanged(sender As Object, e As EventArgs)
        Dim wfId, stepId As Int32
        Presenter.getWorkflowAndStepSelected(wfId, stepId)
        SelectStepNode(stepId)

        RaiseEvent SelectedNodeChanged(wfId, stepId, 0)
        RaiseEvent WFTreeRefreshed(stepId)
    End Sub
End Class

File: Zamba/ZambaWeb2013/zamba2010/Zamba.Web2014/Views/UC/WF/Arbol.ascx.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Web.UI;
using System.Web.UI.WebControls;
using Zamba.Core;

public partial class Arbol : System.Web.UI.UserControl, IArbol
{

    /// <summary>
    /// Una vez finalizado el refresco del arbol de wf se eleva este evento para actualizar la grilla
    /// </summary>
    /// <param name="StepId"></param>
    /// <remarks></remarks>
    public event WFTreeRefreshedEventHandler WFTreeRefreshed;
    public delegate void WFTreeRefreshedEventHandler(Int32 StepId);
    public event WFTreeIsEmptyEventHandler WFTreeIsEmpty;
    public delegate void WFTreeIsEmptyEventHandler();
    public event SelectedNodeChangedEventHandler SelectedNodeChanged;
    public delegate void SelectedNodeChangedEventHandler(Int32 WFId, Int32 StepId, Int32 DocTypeId);

    private Presenters.Arbol Presenter;
    public Arbol()
    {
        Load += Page_Load;
        Init += Page_Init;
    }
    protected void Page_Init(object sender, System.EventArgs e)
    {
        if ((Session["User"] != null))
        {
            Presenter = new Presenters.Arbol(int.Parse(Session["UserId"].ToString()), this, (IUser)Session["User"]);
        }
    }

    protected void Page_Load(object sender, EventArgs e)
    {
        if ((Session["User"] != null))
        {
            btnInsertar.Visible = UserBusiness.Rights.GetUserRights(Zamba.Core.ObjectTypes.InsertWeb, Zamba.Core.RightsType.View);
            if (btnInsertar.Visible)
            {
                Presenter.FillCmbFormTypes();
            }
        }

        string script = "$(document).ready(function() { FixWfTree(); AddStepCountHandler(); });";
        ScriptManager.RegisterClientScriptBlock(this, this.GetType(), "WfTreeScripts", script, true);
    }

    public void Refresh()
    {
        try
        {
            int LastWFStepUsed = int.Parse(UserPreferences.getValue("UltimoWFStepUtilizado", Sections.WorkFlow, 0));
            Int32 WFId = default(Int32);
            Int32 stepID = default(Int32);

            if (LastWFStepUsed > 0)
            {
                Presenter.getWorkflowAndStepSelected(ref WFId, ref LastWFStepUsed);
                SelectStepNode(LastWFStepUsed);

                if (ArbolProcesos.Nodes.Count > 0)
                {
                    bool nodeselected = false;

                    foreach (TreeNode nodeMain in ArbolProcesos.Nodes[0].ChildNodes)
                    {
                        if (nodeMain.Value == LastWFStepUsed.ToString())
                        {
                            nodeMain.Select();
                            if (SelectedNodeChanged != null)
                            {
                                SelectedNodeChanged(WFId, LastWFStepUsed, 0);
                            }
                            break; // TODO: might not be correct. Was : Exit For
                        }

                        foreach (TreeNode NodeChild in nodeMain.ChildNodes)
                        {
                            if (NodeChild.Value == LastWFStepUsed.ToString())
                            {
                                nodeMain.Expand();
                                NodeChild.Select();
                                if (SelectedNodeChanged != null)
                                {
                                    SelectedNodeChanged(WFId, LastWFStepUsed, 0);
                                }
                                nodeselected = true;
                                break; // TODO: might not be correct. Was : Exit For
                            }
                        }
                        if (nodeselected)
                            break; // TODO: might not be correct. Was : Exit For
                    }



                }
            }
            else {
                Presenter.getWorkflowAndStepSelected(ref WFId, ref stepID);
                if (SelectedNodeChanged != null)
                {
                    SelectedNodeChanged(WFId, stepID, 0);
                }
            }
        }
        catch (Exception ex)
        {
            ZCore.raiseerror(ex);
        }
    }

    private void RefreshTree()
    {
        try
        {
            Presenter.FillWF();

            int LastWFStepUsed = int.Parse(UserPreferences.getValue("UltimoWFStepUtilizado", Sections.WorkFlow, 0));
            Int32 WFId = default(Int32);

            Presenter.getWorkflowAndStepSelected(ref WFId, ref LastWFStepUsed);

            if (ArbolProcesos.Nodes.Count > 0)
            {
                SelectStepNode(LastWFStepUsed);
                if (WFTreeRefreshed != null)
                {
                    WFTreeRefreshed(LastWFStepUsed);
                }
            }
        }
        catch (Exception ex)
        {
            ZCore.raiseerror(ex);
        }
    }

    private void SelectStepNode(Int64 stepId)
    {
        ArbolProcesos.Nodes[0].Expand();


        foreach (TreeNode nodeMain in ArbolProcesos.Nodes[0].ChildNodes)
        {
            if (nodeMain.Value == stepId.ToString())
            {
                nodeMain.Select();
            }

            foreach (TreeNode NodeChild in nodeMain.ChildNodes)
            {
                if (NodeChild.Value == stepId.ToString())
                {
                    nodeMain.Expand();
                    NodeChild.Select();
                }
            }
        }
    }

    public void FillWF()
    {
        try
        {
            Presenter.FillWF();

            //Si el arbol no posee nodos, se dispara el evento para que no muestre el control
            if (ArbolProcesos.Nodes.Count == 0)
            {
                if (WFTreeIsEmpty != null)
                {
                    WFTreeIsEmpty();
                }
                return;
            }

            if ((ArbolProcesos.SelectedNode == null))
            {
                Refresh();
            }
        }
        catch (Exception ex)
        {
            ZCore.raiseerror(ex);
        }
    }

    public TreeView WFTreeView
    {
        get { return this.ArbolProcesos; }
        set { this.ArbolProcesos = value; }
    }

    //Protected Sub btncontraer_Click(ByVal sender As Object, ByVal e As System.Web.UI.ImageClickEventArgs) Handles btnContraer.Click
    //    Presenter.CollapseTreeView()
    //End Sub

    public System.Web.UI.WebControls.DropDownList CmbFormType
    {
        get { return this.ddltipodeform; }
        set { this.ddltipodeform = value; }
    }

    protected void BtnRefresh_Click(object sender, EventArgs e)
    {
        RefreshTree();
    }

    protected void ArbolProcesos_SelectedNodeChanged(object sender, EventArgs e)
    {
        Int32 wfId = default(Int32);
        Int32 stepId = default(Int32);
        Presenter.getWorkflowAndStepSelected(ref wfId, ref stepId);
        SelectStepNode(stepId);

        if (SelectedNodeChanged != null)
        {
            SelectedNodeChanged(wfId, stepId, 0);
        }
        if (WFTreeRefreshed != null)
        {
            WFTreeRefreshed(stepId);
        }
    }
  
}